storage/innobase/os/os0file.cc

```diff
@@ -9676,6 +9676,100 @@ bool Encryption::check_keyring()
 	return(false);
 }
 
+/** Encrypt a doublewrite buffer page. The page is encrypted
+using the key of tablespace object provided.
+Caller should allocate buffer for encrypted page
+@param[in]	space			tablespace object
+@param[in]	in_page			unencrypted page
+@param[in,out]	encrypted_buf		buffer to hold the encrypted page
+@param[in]	encrypted_buf_len	length of the encrypted buffer
+@return true on success, false on failure */
+bool
+os_dblwr_encrypt_page(
+	fil_space_t*		space,
+	page_t*			in_page,
+	page_t*			encrypted_buf,
+	ulint			encrypted_buf_len)
+{
+	if (!FSP_FLAGS_GET_ENCRYPTION(space->flags)) {
+		return(false);
+	}
+
+	IORequest	write_request(IORequest::WRITE);
+	page_size_t	page_size(space->flags);
+
+	write_request.encryption_key(
+			space->encryption_key,
+			space->encryption_klen,
+			space->encryption_iv);
+	write_request.encryption_algorithm(
+		Encryption::AES);
+
+	ulint 	bytes = page_size.physical();
+
+	/* After successful encryption, in_page will point
+	to a new memory block which is encrypted and
+	the bytes will have value of length of encrypted data */
+	void*		in_page_before	= in_page;
+	Block* block = os_file_encrypt_page(write_request,
+			in_page_before,
+			&bytes);
```

Comment:
Can os_file_encrypt_page return NULL?

Reply:
The blocks are allocated from block cache (see os_alloc_block()). There is retry logic to get from cache. After this, it allocates a new block. ut_malloc_nokey() which again has retry logic.
Only in very rare scenario, I see block allocation fails.

Do you want me to add assert ?

Reply:
So in theory it's possible to return NULL, although no-one ever is likely to see one. Since it's possible, I'd prefer a check, but in this case ut_ad would be fine I guess

Reply:
Done. added debug assert